File: 1331. rank-transform-of-an-array.java
import java.util.Arrays;
import java.util.HashMap;
class Solution {
    public int[] arrayRankTransform(int[] arr) {
        int[] sorted = arr.clone();

        Arrays.sort(sorted);

        HashMap<Integer, Integer> rank = new HashMap<>();
        int currentRank = 1;

        for (int num : sorted) {
            if (!rank.containsKey(num)) {
                rank.put(num, currentRank++);
            }
        }

        for (int i = 0; i < arr.length; i++) {
            arr[i] = rank.get(arr[i]);
        }

        return arr;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        // Example 1
        int[] arr1 = {40, 10, 20, 30};
        System.out.println(Arrays.toString(sol.arrayRankTransform(arr1))); // [4,1,2,3]

        // Example 2
        int[] arr2 = {100, 100, 100};
        System.out.println(Arrays.toString(sol.arrayRankTransform(arr2))); // [1,1,1]

        // Example 3
        int[] arr3 = {37, 12, 28, 9, 100, 56, 80, 5, 12};
        System.out.println(Arrays.toString(sol.arrayRankTransform(arr3))); // [5,3,4,2,8,6,7,1,3]
    }
}

File: 1331. rank-transform-of-an-array.py
from typing import List

class Solution:
    def arrayRankTransform(self, arr: List[int]) -> List[int]:
        sorted_arr = sorted(arr)

        rank = {}
        current_rank = 1

        for num in sorted_arr:
            if num not in rank:
                rank[num] = current_rank
                current_rank += 1

        return [rank[num] for num in arr]


# Test cases
sol = Solution()

# Example 1
arr = [40, 10, 20, 30]
print(sol.arrayRankTransform(arr))  # [4, 1, 2, 3]

# Example 2
arr = [100, 100, 100]
print(sol.arrayRankTransform(arr))  # [1, 1, 1]

# Example 3
arr = [37, 12, 28, 9, 100, 56, 80, 5, 12]
print(sol.arrayRankTransform(arr))  # [5, 3, 4, 2, 8, 6, 7, 1, 3]